stan/math/prim/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,227 @@
+     *   and accumulate those in member variable sum_. This function may
+     *   be called multiple times per object instantiation (so the sum_
+     *   must be accumulated, not just assigned).
+     *
+     * @param r Range over which to compute `ReduceFunction`
+     */
+    void operator()(const tbb::blocked_range<size_t>& r) {
+      if (r.empty()) {
+        return;
+      }
+
+      std::decay_t<Vec> sub_slice;
+      sub_slice.reserve(r.size());
+      for (int i = r.begin(); i < r.end(); ++i) {
+        sub_slice.emplace_back(vmapped_[i]);
+      }
+
+      sum_ += apply(
+          [&](auto&&... args) {
+            return ReduceFunction()(r.begin(), r.end() - 1, sub_slice, msgs_,
+                                    args...);
+          },
+          this->args_tuple_);
+    }
+
+    /**
+     * Join reducers. Accumuluate the value (sum_) of the other reducer.
+     *
+     * @param rhs Another partial sum
+     */
+    void join(const recursive_reducer& child) { this->sum_ += child.sum_; }
+  };
+
+  /**
+   * Call an instance of the function `ReduceFunction` on every element
+   *   of an input sequence and sum these terms.
+   *
+   * This specialization is parallelized using tbb and works only for
+   *   arithmetic types.
+   *
+   * An instance, f, of `ReduceFunction` should have the signature:
+   *   double f(int start, int end, Vec&& vmapped_subset, std::ostream* msgs,
+   * Args&&... args)
+   *
+   * `ReduceFunction` must be default constructible without any arguments
+   *
+   * Each call to `ReduceFunction` is responsible for computing the
+   *   start through end - 1 terms of the overall sum. All args are passed
+   *   from this function through to the `ReduceFunction` instances.
+   *   However, only elements start through end - 1 of the vmapped argument are
+   *   passed to the `ReduceFunction` instances (as the `vmapped_subset`
+   * argument).
+   *
+   * This function distributes computation of the desired sum
+   *   over multiple threads by coordinating calls to `ReduceFunction`
+   * instances.
+   *
+   * @param vmapped Sliced arguments used only in some sum terms
+   * @param grainsize Suggested grainsize for tbb
+   * @param[in, out] msgs The print stream for warning messages
+   * @param args Shared arguments used in every sum term
+   * @return Summation of all terms
+   */
+  ReturnType operator()(Vec&& vmapped, std::size_t grainsize,
+                        std::ostream* msgs, Args&&... args) const {
+    const std::size_t num_jobs = vmapped.size();
+    if (num_jobs == 0) {
+      return 0.0;
+    }
+    recursive_reducer worker(std::forward<Vec>(vmapped), msgs,
+                             std::forward<Args>(args)...);
+
+#ifdef STAN_DETERMINISTIC
+    tbb::static_partitioner partitioner;
+    tbb::parallel_deterministic_reduce(
+        tbb::blocked_range<std::size_t>(0, num_jobs, grainsize), worker,
+        partitioner);
+#else
```

Comment:
Do we have tests for deterministic mode? We can just run the same test flipping on and off this macro

Reply:
We don't have tests yet. We should probably write test files which just include their partner test files, but do define first the DETERMINSITC macro. So

a_test.cpp

a_deterministic_test.cpp is define of the macro and then an include of a_test.cpp

BTW:

I would suggested we take  the simple_partitioner. That one is the default for parallel_determinstic_reduce and has the property to give you only exactly reproducible results whenever you go with the same number of cores from run to run. The static partitioner is fully deterministic even if you vary the number of cores. However, the simple thing is likely faster and I think its sufficient to be exactly reproducible if the same number of cores is used.

Reply:
> We can just run the same test flipping on and off this macro

Probably just easier to break this out into another function if we need to keep it. ```reduce_sum_deterministic```?

Reply:
Another function is fine, yes.

(I would have thought an if-def is just fine and code-efficient, but ok you both seem to strongly think different here than me).

Reply:
The goal of this is so cause some users might want deterministic code right? And this guarantees that? So we can just have a reduce_sum_deterministic?

Reply:
Yes, that's right. We should actually test the determinism on one unit test at least. We still need to be careful with what we promise here though. Things can still depend on compiler flags... but using the TBB `parallel_deterministic_reduce` is your only hope for to get the same out if you put the same in.

I would be very much in favour for such a function (or compile time flag... whatever).

Reply:
It shall be so then. Easy enough to test. If we can't guarantee exactly deterministic, maybe we should give it another name that implies the same thing but doesn't guarantee it?

```reduce_sum_static```, ```reduce_sum_safe_mode```, or something else?

Reply:
I think it's fine to go with deterministic as that's what we use from the TBB (https://software.intel.com/en-us/node/506155). We just need to explain in the doc.